Xamarin.Forms.Core/Binding.cs

```diff
@@ -113,23 +116,80 @@ internal override void Apply(bool fromTarget)
 			_expression.Apply(fromTarget);
 		}
 
-		internal override void Apply(object context, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false)
+		internal override void Apply(object context, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false, bool fromAncestorChanged = false)
 		{
 			object src = _source;
 			var isApplied = IsApplied;
 
 			base.Apply(src ?? context, bindObj, targetProperty, fromBindingContextChanged: fromBindingContextChanged);
 
-			if (src != null && isApplied && fromBindingContextChanged)
+			if (isApplied && !ShouldReapply(fromBindingContextChanged, fromAncestorChanged))
 				return;
 
-			object bindingContext = src ?? Context ?? context;
+			if ( this.Source is RelativeBindingSource )
+				ResolveRelativeSource(bindObj, targetProperty);
+			else
+				CompleteApplyBinding(src ?? Context ?? context, bindObj, targetProperty);
+		}
+
+		private void CompleteApplyBinding(object bindingContext, BindableObject bindObj, BindableProperty targetProperty)
+		{
 			if (_expression == null && bindingContext != null)
 				_expression = new BindingExpression(this, SelfPath);
-
 			_expression.Apply(bindingContext, bindObj, targetProperty);
 		}
 
+		private async void ResolveRelativeSource(BindableObject bindObj, BindableProperty targetProperty)
+		{
+			RelativeBindingSource relSource = this.Source as RelativeBindingSource;
+			object bindingContext = null;
+			switch (relSource.Mode)
+			{
+				case RelativeBindingSourceMode.Self:
+					bindingContext = bindObj;
+					break;
+				case RelativeBindingSourceMode.TemplatedParent:
+					{
+						var view = bindObj as Element;
+						if (view == null)
+							throw new InvalidOperationException();
+						bindingContext = await TemplateUtilities.FindTemplatedParentAsync(view);
+						break;
+					}
+				case RelativeBindingSourceMode.FindAncestor:
+					{
+						if (!(bindObj is Element elem))
+							throw new InvalidOperationException();
+						Element parent = await TemplateUtilities.GetRealParentAsync(elem);
+						int currentLevel = 1;
+
+						while (parent != null)
+						{
+							if (currentLevel >= relSource.AncestorLevel)
+							{
+								if (relSource.AncestorType.IsInstanceOfType(parent))
+								{
+									bindingContext = parent;
+									break;
+								}
+								else if (relSource.AncestorType.IsInstanceOfType(parent.BindingContext))
```

Comment:
is this how it behaves in other platforms ? I understand it's convenient, but is it the least surprising way ?

Reply:
Good question, and no it's not. This would be an enhancement over WPF/UWP. But I too was starting to have second thoughts about the "surprising" nature of it and so was thinking maybe adding another type of 'RelativeBindingSourceMode' in its own right, e.g., `FindAncestorBindingContext`. Then there's no ambiguity that you're looking to match by the parent binding context rather than the parent's view type. Would that be better?